src/main/scala/mesosphere/marathon/state/Group.scala

```diff
@@ -152,13 +153,64 @@ object Group extends StrictLogging {
   def validGroup(base: PathId, enabledFeatures: Set[String]): Validator[Group] =
     validator[Group] { group =>
       group.id is validPathWithBase(base)
-      group.apps.values as "apps" is every(
-        AppDefinition.validNestedAppDefinition(group.id.canonicalPath(base), enabledFeatures))
+
+      group.transitiveApps as "apps" is everyApp(
+        AppDefinition.validBasicAppDefinition(enabledFeatures) and isChildOfParentId(group)
+      )
+
       group is noAppsAndPodsWithSameId
       group is noAppsAndGroupsWithSameName
       group is noPodsAndGroupsWithSameName
-      group.groupsById.values as "groups" is every(validGroup(group.id.canonicalPath(base), enabledFeatures))
+
+      group.transitiveGroupValues as "groups" is every(
+        noAppsAndPodsWithSameId and noAppsAndGroupsWithSameName and noPodsAndGroupsWithSameName
+          and isTrue("Group has to be child of groups with parent id") { childGroup =>
+            if (childGroup.id.parent == group.id) group.groupsById.contains(childGroup.id)
```

Comment:
oh, I get it now... this row is just an optimization to not do lookout when we are dealing with parent...

that do you think about moving this optimization directly to the group method (it can be it's implementation detail). Then we can take advantage of this speedup everywhere...